### ISSUE TITLE
Analysis: Neo4j query efficiency

### ISSUE THREAD

**Issue by shirey** [hubmapconsortium/entity-api#517]
Are the Neo4j queries as efficient as they can be?

Extract all of the Neo4j queries for the GET of a Publication that has many direct ancestor datasets (> 100).  Run the queries, recording the execution times and analyze to see if the queries can be restructured, have added information (less [*] relationships, possibly?) or even have Neo4j indices added to improved the performance.

**Comment by yuanzhou** [comment]
Sunset also reported a timeout issue:
>Not sure if this is related to the FS issue we currently have, but I'm also facing timeouts with entity-api (https://entity.api.hubmapconsortium.org/datasets/unpublished?format=json) I know this should be a big query.

Essentially it's a neo4j cypher with `OPTIONAL MATCH` which is causing performance issue thus timed out.

**Comment by DerekFurstPitt** [comment]
Yep that was my suspicion as well. The fix should be reasonably simple. If I can't find a way to restructure the original query, worst case scenario, splitting into 2 queries should drastically speed things up.

**Comment by yuanzhou** [comment]
@DerekFurstPitt as suggested by Bill, this card is separate from the timeout issue (although another slow Neo4j query) that Sunset report. And I created a separate card for it that we can visit later: https://github.com/hubmapconsortium/entity-api/issues/518

**Comment by DerekFurstPitt** [comment]
The following queries are executed during the processes of the `GET` request on a Publication

**schema_neo4j_queries.get_entity(neo4j_driver, uuid)**
MATCH (e:Entity) 
WHERE e.uuid = '{uuid}' 
RETURN e AS {record_field_name}

**Schema_neo4j_queries.get_dataset_collections(neo4j_driver, uuid, property_key)**
MATCH (e:Entity)-[:IN_COLLECTION]->(c:Collection)
WHERE e.uuid = '{uuid}'
RETURN apoc.coll.toSet(COLLECT(c)) AS {record_field_name}

**Schema_neo4j_queries.get_dataset_uploads(neo4j_driver, uuid, property_key)**
MATCH (e:Entity)-[:IN_UPLOAD]->(s:Upload)
WHERE e.uuid = '{uuid}'
RETURN s AS {record_field_name}

**Schema_neo4j_queries.get_dataset_direct_ancestors(neo4j_driver, uuid, property_key)**
MATCH (s:Entity)-[:ACTIVITY_INPUT]->(a:Activity)-[:ACTIVITY_OUTPUT]->(t:Dataset)
WHERE t.uuid = '{uuid}' "
RETURN apoc.coll.toSet(COLLECT(s)) AS {record_field_name}

**Schema_neo4j_queries.get_previous_revision_uuid(neo4j_driver, uuid)**
MATCH (e:Entity)-[:REVISION_OF]->(previous_revision:Entity)
WHERE e.uuid='{uuid}'
RETURN previous_revision.uuid AS {record_field_name}

**Schema_neo4j_queries.get_next_revision_uuid(neo4j_driver, uuid)**
MATCH (e:Entity)<-[:REVISION_OF]-(next_revision:Entity)
WHERE e.uuid='{uuid}' "
RETURN next_revision.uuid AS {record_field_name}

**Schema_neo4j_queries.get_publication_associated_collection(neo4j_driver, uuid)**
MATCH (p:Publication)-[:USES_DATA]->(c:Collection)
WHERE p.uuid = '{uuid}'
RETURN c as {record_field_name}

These are all very simple queries. No optional matches, no wildcard or variable length relationships. All nodes have explicit labels. And they are all keyed for specific UUID's. The only one with the possibility of returning very many nodes is get_direct_ancestors and it uses apoc.coll.toSet to efficiently put all the unique nodes into a list. I see no obvious optimizations to be made. The problem doesn't appear to be on the query side of things. The timeout issue appears it may have more to do with indexing and factors outside of the queries themselves. 

There are certainly optimizations that can be made for our various endpoints app_neo4j_queries, but the schema_neo4j_queries used with our trigger methods are all very efficient.
